pkg/store/consul.go

```diff
@@ -91,11 +125,47 @@ func (s *Consul) Get(key string) (*KVPair, error) {
 }
 
 // Put a value at "key"
-func (s *Consul) Put(key string, value []byte) error {
-	p := &api.KVPair{Key: s.normalize(key), Value: value}
-	if s.client == nil {
-		log.Error("Error initializing client")
+func (s *Consul) Put(key string, value []byte, opts *WriteOptions) error {
+
+	key = s.normalize(key)
+
+	p := &api.KVPair{
+		Key:   key,
+		Value: value,
 	}
+
+	if opts != nil && opts.Ephemeral {
+		// Creates the global ephemeral session
+		// if it does not exist
+		if s.ephemeralSession == "" {
+			s.createEphemeralSession()
+		}
+
+		// Create lock option with the
+		// EphemeralSession
+		lockOpts := &api.LockOptions{
```

Comment:
After tinkering a little bit with Consul and the lock mechanism, if we give the same session to the `api.KVPair` and `api.LockOptions`, the call to `lock.Lock()` is non blocking because of those lines in `consul/lock.go`:

```
if pair != nil && pair.Session == l.lockSession {
		goto HELD
}
```
I don't think we can put it in `createEphemeralSession` either, we need to go through the lock call each time just in case the session expired and we want to come back to the discovery service if we recover (but I need to add one line during an error of refresh if the session does not exists anymore and put it back to `""`).

Also, I thought that this might leak goroutines by just calling `lock.Lock()` each time because of this line in `consul/lock.go`:

```
go l.monitorLock(l.lockSession, leaderCh)
```
But it appears that no :D

It's just a placeholder lock that does nothing else but hold the ephemeral session :)